f8a_worker/workers/graph_populator.py

```diff
@@ -0,0 +1,242 @@
+import logging
+import re
+import time
+from datetime import datetime
+
+logger = logging.getLogger(__name__)
+
+# Compile patterns to clean a string (used for safety against GraphQL injection attack)
+danger_word_list = ["drop\(\)", "V\(\)", "count\(\)"]
+remove = '|'.join(danger_word_list)
+pattern = re.compile(r'(' + remove + ')', re.IGNORECASE)
+pattern_to_save = '[^\w\*\.Xx\-\>\=\<\~\^\|\/\:]'
+pattern_n2_remove = re.compile(pattern_to_save)
+
+
+class GraphPopulator(object):
+    @classmethod
+    def str_value_cleaner(cls, safe_string):
+        """
+        Clean up the safe_string to remove suspicious words,
+        and special characters.
+        """
+        result = ''
+        if isinstance(safe_string, str):  # can raise an exception as well
+            temp_str = pattern.sub('', safe_string)
+            result = pattern_n2_remove.sub('', temp_str)
+        return result
+
+    @classmethod
+    def construct_version_query(cls, input_json):
+        description = ''
+        try:
+            if len(input_json.get('analyses', {}).get('metadata', {}).get('details')) > 0:
+                description = input_json.get('analyses').get('metadata').get('details')[0].get('description', '')
+                description = cls.str_value_cleaner(description).replace("'", "\\'")
+        except:
+            # we pass and move forward without description
+            pass
+
+        drop_props = []
+        str_version = prp_version = drop_prop = ""
+        # Check if license and cve analyses succeeded. Then we refresh the property
+        if 'success' == input_json.get('analyses', {}).get('source_licenses', {}).get('status'):
+            drop_props.append('licenses')
+        if 'success' == input_json.get('analyses', {}).get('security_issues', {}).get('status'):
+            drop_props.append('cve_ids')
+        if len(drop_props) > 0:
+            drop_prop += "g.V().has('pecosystem', ecosystem).has('pname', pkg_name)" \
+                         ".has('version', version).properties('" + "','".join(drop_props) + "')." \
+                         "drop().iterate();" \
+
+        str_version += "ver = g.V().has('pecosystem', ecosystem).has('pname', pkg_name)" \
+                       ".has('version', version).tryNext().orElseGet{graph.addVertex('pecosystem', ecosystem," \
+                       "'pname', pkg_name, 'version', version, " \
+                       "'vertex_label', 'Version')};" \
+                       "ver.property('last_updated'," + str(time.time()) + ");"
+        # Add Description if not blank
+        if description:
+            prp_version += "ver.property('description','" + re.sub('[^A-Za-z0-9_\\\/\'":. ]', '', description) + "');"
+        # Get Code Metrics Details
+        if 'code_metrics' in input_json.get('analyses', {}):
```

Comment:
What if there is no `code_metrics`. Is it correct state?

Reply:
if there is no code_metrics collected it will just skip

Reply:
Is it not missing in the graph then?

Reply:
nope. Graph does not require code_metrics as mandatory. Infact, now we do not show code_metrics at all, and we have also disabled in the workflow. @fridex can correct me if I am wrong.

Reply:
[Code metrics is disabled](https://github.com/fabric8-analytics/fabric8-analytics-worker/blob/master/f8a_worker/dispatcher/flows/bayesianAnalysisFlow.yml#L45). In that case please remove this dead code.

Reply:
What are the chances that it will never be enabled?